tests/cases.py

```diff
@@ -108,3 +112,97 @@ def setUp(self):
     def tearDown(self) -> None:
         """Tear down the test case by cleaning up the temporary cache directory."""
         self.directory.cleanup()
+
+
+class LossTestCase(unittest.TestCase):
+    """Base unittest for loss functions."""
+
+    #: The class
+    cls: ClassVar[Type[Loss]]
+
+    #: Constructor keyword arguments
+    kwargs: ClassVar[Optional[Mapping[str, Any]]] = None
+
+    #: The loss instance
+    loss: Loss
+
+    #: The batch size
+    batch_size: ClassVar[int] = 3
+
+    def setUp(self) -> None:
+        """Initialize the instance."""
+        kwargs = self.kwargs
+        if kwargs is None:
+            kwargs = {}
+        self.loss = self.cls(**kwargs)
+
+    def _check_loss_value(self, loss_value: torch.FloatTensor) -> None:
+        """Check loss value dimensionality, and ability for backward."""
+        # test reduction
+        self.assertEqual(0, loss_value.ndim)
+
+        # Test backward
+        loss_value.backward()
+
+
+class PointwiseLossTestCase(LossTestCase):
+    """Base unit test for label-based losses."""
+
+    #: The number of entities.
+    num_entities: int = 17
+
+    def test_type(self):
+        """Test the loss is the right type."""
+        self.assertIsInstance(self.loss, PointwiseLoss)
+
+    def test_label_loss(self):
+        """Test ``forward(logits, labels)``."""
+        logits = torch.rand(self.batch_size, self.num_entities, requires_grad=True)
+        labels = functional.normalize(torch.rand(self.batch_size, self.num_entities, requires_grad=False), p=1, dim=-1)
+        loss_value = self.loss.forward(
+            logits,
+            labels,
+        )
+        self._check_loss_value(loss_value)
+
+
+class PairwiseLossTestCase(LossTestCase):
+    """Base unit test for pair-wise losses."""
+
+    #: The number of negative samples
+    num_negatives: int = 5
+
+    def test_type(self):
+        """Test the loss is the right type."""
+        self.assertIsInstance(self.loss, PairwiseLoss)
+
+    def test_pair_loss(self):
+        """Test ``forward(pos_scores, neg_scores)``."""
+        pos_scores = torch.rand(self.batch_size, 1, requires_grad=True)
```

Comment:
there is some code duplication for these tests.

Also: Do not explicitly call forward, but instead use `loss(...)`

Reply:
yes for this PR i was just reorganizing. next one will update the code itself.